### PR TITLE
Initial work for the params search

### PR DESCRIPTION
**Pull request by akshay23**
img_0359
img_0358
- Can enter new keywords to change songs
- BPM not working yet
- Minor UI fixes
- Warning fixes

### CODE REVIEW

File: SCloudRandomizer/Base.lproj/Main.storyboard [hunk 1]
```diff
@@ -499,7 +499,7 @@
...
                                     <constraint firstItem="WVR-6F-hpQ" firstAttribute="top" secondItem="Gvq-Ge-NtX" secondAttribute="bottom" constant="16" id="cn8-kW-QlF"/>
                                     <constraint firstItem="Nbf-RZ-irD" firstAttribute="leading" secondItem="oFN-SU-Zzn" secondAttribute="leading" constant="21" id="efa-pE-gcX"/>
-                                    <constraint firstItem="Cxj-kG-kty" firstAttribute="top" secondItem="oFN-SU-Zzn" secondAttribute="top" constant="23" id="f1k-ip-Wjh"/>
+                                    <constraint firstItem="Cxj-kG-kty" firstAttribute="top" secondItem="oFN-SU-Zzn" secondAttribute="top" constant="31" id="f1k-ip-Wjh"/>
                                     <constraint firstItem="mCp-sK-lzn" firstAttribute="trailing" secondItem="WVR-6F-hpQ" secondAttribute="trailing" id="gXy-1D-rjc"/>
                                     <constraint firstItem="Nbf-RZ-irD" firstAttribute="centerX" secondItem="eVh-xf-eve" secondAttribute="centerX" id="h81-Y4-EU4"/>
```
Comment (VegaFromLyra) [inline SCloudRandomizer/Base.lproj/Main.storyboard:504]: This is going to be a shitty merge conflict :P

File: SCloudRandomizer/MainViewController.m [hunk 2]
```diff
@@ -35,9 +36,10 @@ - (void)viewDidLoad
     
     // New search params view instance and track info VC
     self.searchParamsVC = [[GlobalData getInstance].mainStoryboard instantiateViewControllerWithIdentifier:@"searchParamsVC"];
+    self.searchParamsVC.delegate = self;
     self.trackInfoVC = [[GlobalData getInstance].mainStoryboard instantiateViewControllerWithIdentifier:@"trackInfoVC"];
     self.trackInfoVC.delegate = self;
-    self.mySearchParams = [[SearchParams alloc] initWithBool:YES];
+    self.mySearchParams = [[SearchParams alloc] initWithBool:YES keywords:@"Biggie,2pac,remix" lowBpm:[NSNumber numberWithInt:80] highBpm:[NSNumber numberWithInt:150]];
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m:43]: [nit] you could put the parameters on separate lines so its easier to read

File: SCloudRandomizer/MainViewController.m [hunk 3]
```diff
@@ -227,7 +239,7 @@ - (void)playPauseSong
     }
 }
 
-// TODO: Change logic to get tracks based on SearchParams object
+// Get tracks based on SearchParams
 - (void)getTracks
 {
     SCRequestResponseHandler handler;
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m]: Can move declaration and initialization of 'handler' to the same line

File: SCloudRandomizer/MainViewController.m [hunk 4]
```diff
@@ -227,7 +239,7 @@ - (void)playPauseSong
     }
 }
 
-// TODO: Change logic to get tracks based on SearchParams object
+// Get tracks based on SearchParams
 - (void)getTracks
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m]: Another way we could do this is pass a url to getTracks and ask it to fetch the songs from that url. So that way getTracks does not need to know which url it needs to use, we can pass that responsibility to the caller (plus its more testable that way!)

Also we could
1) On first load, fetch the favorite tracks using this method instead of hardcoding the search term. (if there are no fav tracks, we could pop up another modal asking them to explicitly set a search term)
2) Construct url based on the search term and then use this method whenever user changes search term

File: SCloudRandomizer/MainViewController.m [hunk 5]
```diff
@@ -65,7 +67,17 @@ - (void)viewWillAppear:(BOOL)animated
         
         if (self.mySearchParams.hasParamsChanged)
         {
-            [self getTracks];
+            
+            if ([self.player isPlaying])
+            {
+                [self refreshTrackList];
+                [self doPlayNextSong];
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m:73]: doPlayNextSong already seems to be calling refereshTrackList so we probably don't need it here?

File: SCloudRandomizer/MainViewController.m [hunk 6]
```diff
@@ -354,7 +369,7 @@ - (void)getTrackInfo:(NSDictionary *)track shouldPlay:(BOOL)play
 
 }
 
-// TODO: Load songs based on SearchParams
+// Load songs based on SearchParams
 // Refresh/reload track list
 - (void)refreshTrackList
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m]: Same comment as before, we can update this to take a url in as well. Lets discuss it!!

File: SCloudRandomizer/SearchParams.h [hunk 7]
```diff
@@ -7,11 +7,15 @@
 //
 
 #import <Foundation/Foundation.h>
+#import "GlobalData.h"
```
Comment (VegaFromLyra) [inline SCloudRandomizer/SearchParams.h:10]: Why do we need GlobalData here?

Comment (VegaFromLyra) [inline SCloudRandomizer/SearchParams.h:10]: oh I think you meant to add it to SearchParamsVC.h? Because its being used in SearchParamsVC.m

File: SCloudRandomizer/SearchParamsVC.m [hunk 8]
```diff
@@ -38,6 +54,23 @@ - (void)didReceiveMemoryWarning {
 - (IBAction)done:(id)sender
 {
     [self dismissKeyboard];
+    
+    if (![GlobalData stringIsNilOrEmpty:self.txtKeywords.text])
+    {
+        NSNumberFormatter *f = [[NSNumberFormatter alloc] init];
```
Comment (VegaFromLyra) [inline SCloudRandomizer/SearchParamsVC.m]: what the f! :P Longer name please :D

File: SCloudRandomizer/SearchParams.h (hunk 7 above)
Comment (akshay23) [inline SCloudRandomizer/SearchParams.h:10]: I need GlobalData here so I can use one of the helper functions in it to compare entries in the future (ie. null or empty bpm entries)

File: SCloudRandomizer/Base.lproj/Main.storyboard (hunk 1 above)
Comment (akshay23) [inline SCloudRandomizer/Base.lproj/Main.storyboard:504]: I will undo the storyboard changes

File: SCloudRandomizer/SearchParamsVC.m (hunk 8 above)
Comment (akshay23) [inline SCloudRandomizer/SearchParamsVC.m]: done

File: SCloudRandomizer/MainViewController.m [hunk 9]
```diff
@@ -227,29 +238,47 @@ - (void)playPauseSong
...
 }
 
-// TODO: Change logic to get tracks based on SearchParams object
-- (void)getTracks
+// Get/refresh tracks based on SearchParams
+- (void)getTracks:(BOOL)shouldGetTrackInfo shouldPlay:(BOOL)playBool
 {
-    SCRequestResponseHandler handler;
-    handler = ^(NSURLResponse *response, NSData *data, NSError *error) {
+    // Disable buttons
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m:244]: [nit] comment should be on next line

File: SCloudRandomizer/MainViewController.m [hunk 10]
```diff
@@ -227,29 +238,47 @@ - (void)playPauseSong
...
+    self.imgArtwork.alpha = 0.5;
+    [self.btnPlay setHidden:NO];
+    [self.btnNext setHidden:NO];
+    [self.btnNext setEnabled:NO];
+    [self.btnPlay setEnabled:NO];
+    [self.btnInfo setEnabled:NO];
+    [self.btnLike setEnabled:NO];
+    [self.btnChangeParams setEnabled:NO];
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m:252]: We don't need to setHidden on btnInfo, btnLike and btnChangeParams?

File: SCloudRandomizer/MainViewController.m [hunk 11]
```diff
@@ -310,86 +309,98 @@ - (void)getTrackInfo:(NSDictionary *)track shouldPlay:(BOOL)play
     NSString *streamURL = [track objectForKey:@"stream_url"];
     NSLog(@"The streamURL is: %@", streamURL);
     
+    // Get new track if stream URL is nil
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m:312]: I am a little confused as to what this while block is doing and why we have it this way but lets discuss that in person

File: SCloudRandomizer/MainViewController.m [hunk 12]
```diff
@@ -310,86 +309,98 @@ - (void)getTrackInfo:(NSDictionary *)track shouldPlay:(BOOL)play
...
     }
     
+    // Progress/loading hud
     SCAccount *account = [SCSoundCloud account];
     MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:self.view animated:YES];
     hud.mode = MBProgressHUDModeIndeterminate;
     hud.labelText = @"Loading Track";
     hud.detailsLabelText = @"Please wait..";
     
+    // Progress handler
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m:328]: We can move out the progress handler logic into its own class and have the main view controller use it when it needs. We can do this in a separate PR

File: SCloudRandomizer/MainViewController.m [hunk 13]
```diff
@@ -310,86 +309,98 @@ - (void)getTrackInfo:(NSDictionary *)track shouldPlay:(BOOL)play
...
+    SCRequestSendingProgressHandler progressHandler;
+    progressHandler = ^(unsigned long long bytesSent, unsigned long long bytesTotal) {
+        hud.progress = ((float)bytesSent / bytesTotal);
+    };
+    
+    // Response handler
+    SCRequestResponseHandler responseHandler;
```
Comment (VegaFromLyra) [inline SCloudRandomizer/MainViewController.m:335]: you can do declaration and initialization on the same line

File: SCloudRandomizer/SearchParams.h (hunk 7 above)
Comment (VegaFromLyra) [inline SCloudRandomizer/SearchParams.h:10]: hmm I don't see any helper functions being added to SearchParams.m tho?

**Comment by VegaFromLyra** [comment]
:dancers:  There's some more clean up I think we can do but lets do that in another PR

File: SCloudRandomizer/MainViewController.m [hunk 14]
```diff
@@ -227,29 +238,47 @@ - (void)playPauseSong
...
+    self.imgArtwork.alpha = 0.5;
+    [self.btnPlay setHidden:NO];
+    [self.btnNext setHidden:NO];
+    [self.btnNext setEnabled:NO];
+    [self.btnPlay setEnabled:NO];
+    [self.btnInfo setEnabled:NO];
+    [self.btnLike setEnabled:NO];
+    [self.btnChangeParams setEnabled:NO];
+    
+    // Progress/loading hud
```
Comment (akshay23) [inline SCloudRandomizer/MainViewController.m:254]: Create common method to fetch progress bar
